File: Segundo_Curso/Segundo_Cuatri/PDOO/ruby/Irrgarten/Player.rb
#encoding: UTF-8

require_relative 'Dice'
require_relative 'Weapon'
require_relative 'Shield'
require_relative 'Directions'
require_relative 'LabyrinthCharacter'

module Irrgarten

    class Player < LabyrinthCharacter
        @@DEFAULT_NAME = "Player #"

        @@MAX_WEAPONS = 2
        @@MAX_SHIELDS = 3
        @@INITIAL_HEALTH = 10
        @@HITS2LOSE = 3

        # Inicializador
        # number -> Numero de jugador
        # intelligence -> Inteligencia del jugador
        # strength -> Fuerza del jugador
        def initialize(number,intelligence,strength)
            @number = number
            super(@@DEFAULT_NAME+number.to_s,intelligence,strength,@@INITIAL_HEALTH)
            @consecutiveHits = 0
            @weapons = Array.new        # Declara un array de objetos, tambien vale []
            @shields = Array.new
        end

        # Constructor de copia de un Player
        # other -> jugador del que copiar
        def cloner(other)
            super
            @number = other.number
            @consecutiveHits = other.consecutiveHits

            #Esto provoca que tengan el mismo array de elementos de combate no una copia.
            #@weapons = other.weapons
            #@shields = other.shields
            @weapons = Array.new
            @shields = Array.new
            copy(other.weapons,@weapons)
            copy(other.shields,@shields) #TODO: revisar la funcion copy

        end

        # Resucita a un jugador
	    def resurrect
            @weapons.clear()
            @shields.clear()
            @health = @@INITIAL_HEALTH
            @consecutiveHits = 0
        end

        #getter row, col, number
        attr_reader :row
        attr_reader :col
        attr_reader :number
        attr_reader :weapons
        attr_reader :shields
        attr_reader :consecutiveHits

        # Modificador de la posicion de un jugador
        # row -> fila nueva
        # col -> columna nueva
        # @pre la posicion debe estar dentro del tablero
        def setPos(row,col)
            if row >= 0 && col >= 0 then
                @row = row
                @col = col
            end
        end

        # Comprueba si el jugador esta muerto
        def dead
            @health <= 0
        end

        # Mueve al jugador a la posicion indicada por direction si esta en validMoves
        # direction -> direccion de movimiento
        # validMoves -> movimientos validos
        # si el movimiento realizado no esta en los validos se mueve uno aleatorio dentro de los validos
	    def move(direction, validMoves)
            size = validMoves.length
            contained = find(direction,validMoves)      # existe la funcion include?()
            if (size > 0) && !contained then
                firstElement = validMoves[0]
                return firstElement
            else
                return direction
            end
        end

        # Calcula el ataque de un jugador
	    def attack
            @strength + sumWeapons
        end

        # Calcula la defensa de un jugador y maneja la recepcion de un golpe.
	    def defend(receivedAttack)
            manageHit(receivedAttack)
        end

        # Asignar los recursos que obtiene el jugador en una pelea si gana
        # Puede añadir: vida, armas y escudos
        def receiveReward
            wReward = Dice.weaponsReward
            sReward = Dice.shieldsReward

            for i in 0..wReward do
                wnew = newWeapon
                receiveWeapon(wnew)
            end
            for i in 0..sReward do
                snew = newShield
                receiveShield(snew)
            end
            extraHealth = Dice.healthReward
            @health += extraHealth
        end

        # Concatena la informacion de un Player
        # Cadena con la informacion
	    def to_s
            str="P" + super
            str+= "\t\tWeapons: ["
            str += @weapons[0].to_s unless @weapons.size==0

            for i in 1...@weapons.length do         # for element in @weapons
                str += ", " + @weapons[i].to_s
            end
            str+= "]\n"
            str+= "\t\tShields: ["
            str += @shields[0].to_s unless @shields.size==0

            for i in 1...@shields.size do
                str += ", " + @shields[i].to_s
            end
            str += "]\n"
            str

        end

	private

        # Encargado de obtener la probabilidad de recibir un nuevo arma y borra las que tomen probabilidad de borrado igual a true (wi.discard())
        # Añade un arma si hay huecos en la mochila del jugador
        def receiveWeapon(w)
            i=0
            while i<@weapons.size
                wi=@weapons[i]
                if(wi.discard()) then
                        @weapons.shift      # Extrae los n primeros elementos del array, sin args devuelve el primero
                else
                    i+=1
                end
            end
            size = @weapons.size
            if size<@@MAX_WEAPONS then
                w = newWeapon
                @weapons.append(w)          # Añade un elemento al final del array, similar a <<
            end
        end
        # Encargado de obtener la probabilidad de recibir un nuevo escudo y borra las que tomen probabilidad de borrado igual a true (wi.discard())
        # Añade un escudo si hay huecos en la mochila del jugador
        def receiveShield(s)
            i=0
            while i<@shields.length
                si=@shields[i]
                if(si.discard) then
                    @shields.shift
                else
		            i+=1
		    end
	    end
            size = @shields.length
            if(size<@@MAX_SHIELDS) then
                s = newShield
                @shields.append(s)
            end
        end

        # Gesstiona los golpes que recibe el jugador
        # receivedAttack -> valor del ataque recibido
        # return -> lose
        #               True -> Jugador muere
        #               False -> No muere
        def manageHit(receivedAttack)
            defense=defensiveEnergy
            if defense<receivedAttack then
                gotWounded
                incConsecutiveHits
            else
                resetHits
            end
            lose=true
            if (@consecutiveHits == @@HITS2LOSE)||(dead()) then
                resetHits
            else
                lose=false
            end
            return lose
        end

        # Crea un nuevo arma y lo devuelve
        def newWeapon
            Weapon.new(Dice.weaponPower, Dice.usesLeft)
        end

        # Crea un nuevo escudo y lo devuelve
        def newShield
            Shield.new(Dice.shieldPower, Dice.usesLeft)
        end

        # Resetea los hits
        def resetHits
            @consecutiveHits = 0
        end

        # Decrementa el valor de health
        def gotWounded
            @health -= 1
        end

        # Incrementa los hits consecutivos
        def incConsecutiveHits
            @consecutiveHits +=1
        end

    protected
        # Suma y devuelve el valor de ataque de todas las armas
        def sumWeapons
            sum=0
            for i in 0...@weapons.size do   #rango exclusivo
                sum += @weapons[i].attack
            end
            sum
        end

        # Suma y devuelve el valor de defensa de todos los escudos
        def sumShields
            sum=0
            for i in 0..@shields.size-1 do  # rango inclusivo
                sum += @shields[i].protect
            end
            sum
        end

        # Devuelve la enegia defensiva
        def defensiveEnergy
            @intelligence + sumShields
        end
    private
        # Busca un elemento en un array
        # Se usa para encontrar una direccion en validMoves
        # element -> Elemento a buscar
        # array -> Array donde buscar
        # True -> Encontrado
        # False -> No encontrado
        def find(element,array)
            found = false
            i = 0
            while !found && (i<array.size)
                if array[i] == element then
                    found = true
                else
                    i += 1
                end
            end
            found
        end

        # Copia un array en otro, ninguna comprobacion de tipos es hecha
        # Se usa para el "constructor" de copia
        # emisor -> vector a copiar
        # receptor -> vector que recibe la copia
        # return -> receptor, vector ya copiado
        def copy(emisor, receptor)
                #Esta manera tambien copia las mismas armas, aunque ahora el array si es diferente
                #Se copia la referencia de cada objeto
            # if !emisor.nil? then
            #     for i in 0...emisor.size do
            #         receptor[i]=emisor[i]
            #     end
            # end
            # #TODO: quitar esto que es de debug
            # puts receptor[0].equal? emisor[0]
            # receptor
            #Con el dup duplicas el objeto en si, no la referencia
            if !emisor.nil? && !receptor.nil? then
                receptor.clear()
                emisor.each{|e| receptor << e.dup}
            end
            receptor
        end
    end #class
end #module

File: Segundo_Curso/Segundo_Cuatri/PDOO/java/src/irrgarten/Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;
import java.util.Iterator;
/**
 *
 * @author el_dramas
 */
public class Player extends LabyrinthCharacter {
    static private final String DEFAULT_NAME = "Player #";   
    static private int MAX_WEAPONS = 2;
    static private int MAX_SHIELDS = 3;
    static private int INITIAL_HEALTH = 10;
    static private int HITS2LOSE = 3;
    
    private char number;
    private int consecutiveHits = 0;
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields; 

    private ShieldCardDeck shieldDeck;
    private WeaponCardDeck weaponDeck;
    
    /**
     * Construye un jugador
     * @param number Su numero 
     * @param intelligence su inteligencia / defensa
     * @param strength su fuerza
     */
    public Player(char number, float intelligence, float strength){
        super (DEFAULT_NAME+number,intelligence,strength,INITIAL_HEALTH);
        weapons = new ArrayList<Weapon>();
        shields = new ArrayList<Shield>();

        shieldDeck = new ShieldCardDeck();
        weaponDeck = new WeaponCardDeck();

        this.number = number;
    }
    
    /**
     * Constructor de copia de un jugador
     * @param other Jugador donde el que copiar
     */
    public Player(Player other){
        super (other);
        this.consecutiveHits = other.consecutiveHits;
        this.number = other.getNumber();
        this.weapons = new ArrayList<Weapon>();
        for(Weapon w: other.weapons){
            this.weapons.add(new Weapon(w));
        }
        this.shields = new ArrayList<Shield>();
        for(Shield s: other.shields){
            this.shields.add(new Shield(s));
        }

        weaponDeck = new WeaponCardDeck();
        shieldDeck = new ShieldCardDeck();
    }

    /**
     * Maneja la resurreccion
     */
    public void resurrect(){
        this.resetHits();
        this.setHealth(INITIAL_HEALTH);
        this.shields.clear();
        this.weapons.clear();        
    }

    
    /**
     * @return El numero del jugador
     */
    public char getNumber(){
        return number;
    }
    
    /**
     * Comprueba si el jugador esta muerto
     * @return True si esta muerto
     */
    @Override
    public boolean dead(){
        return super.dead();
    }
    
    /**
     * Porporciona una direccion de movimiento de un jugador, si direction es correcta es la que se devuelve.
     * @param direction La direccion preferida
     * @param validMoves Direcciones posibles
     * @return La direccion a la que mover al jugador
     */
    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        Directions output = direction;
        if(!validMoves.isEmpty() && !validMoves.contains(direction)){   //Comprueba si la direccion esta en el array
            output = validMoves.get(0);
        }
        return output;
    }
    
    /**
     * Maneja el ataque
     * @return la fuerza del ataque
     */
    @Override
    public float attack(){
        if(this.dead())
            return 0;
        else return super.getStrength()+this.sumWeapons();
    }
    
    /**
     * Maneja el recibo de recompensas
     * 
     */
    public void receiveReward(){
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        while(wReward != 0){
            wReward--;
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }
        while(sReward != 0){
            sReward--;
            Shield snew = newShield();
            receiveShield(snew);
        }
        super.setHealth(super.getHealth()+Dice.healthReward());
    }


    /**
     * Defiende de un ataque
     * @param receivedAttack La fuerza del ataque
     * @return true si ha perdido por el golpe
     */
    @Override
    public boolean defend(float receivedAttack){
        return this.manageHit(receivedAttack);
    }
    
    /**
     * Metodo que concatena el estado
     * @return Cadena con la informacion de estado de la instancia
     */
    @Override
    public String toString(){
        String cad= "P" + super.toString() + "\n";
        
        cad += "Weapons:\n";
        
        for (int i=0; i<weapons.size()-1; ++i){
            cad += weapons.get(i).toString() + " - ";
        }
        
        if(!weapons.isEmpty()){
            cad += weapons.get(weapons.size()-1).toString() + "\n";
        }
        
        cad += "Shields:\n";
        
        for (int i=0; i<shields.size()-1; ++i){
            cad += shields.get(i).toString() + " - ";
        }
        
        if(!shields.isEmpty()){
            cad += shields.get(shields.size()-1).toString() + "\n";
        }
        return cad;
    }
    
    /**
     * Recibe un arma y maneja el inventario
     * @param w el arma
     */
    private void receiveWeapon(Weapon w){

        Iterator<Weapon> it = weapons.iterator(); //Iteradores en java
        while(it.hasNext()){                      //Comprueba si hay otro 
            Weapon wl = it.next();                //Devuelve el siguiente
            if( wl.discard()){
                it.remove();
            }
        }
        
        if(weapons.size() < MAX_WEAPONS){
            weapons.add(w);                        //Añade en la ultima posicion
        }
    }
    
    /**
     * Recibe un escudo y maneja el inventario
     * @param s el escudo
     */
    private void receiveShield(Shield s){
        Iterator<Shield> it = shields.iterator();   //Lo inicializa antes del primero
        
        while(it.hasNext()){
            Shield sl = it.next();  //Devuelve el siguiente
            if( sl.discard()){
                it.remove();
            }
        }

        if(shields.size()<MAX_SHIELDS){
            shields.add(s);
        }
    }
    
    /**
     * Coge un arma de la baraja 
     * @return el arma
     */
    private Weapon newWeapon(){
        return weaponDeck.nextCard();
    }
    
    /**
     * Coge un escudo de la baraja
     * @return el escudo
     */
    private Shield newShield(){
        return shieldDeck.nextCard();
    }
    
    /**
     * Calcula el ataque con tus armas
     * @return el ataque total
     */
    protected float sumWeapons(){
        float fullAttack = 0;
        
        for (int i=0; i<weapons.size(); ++i){
            fullAttack += weapons.get(i).attack();      //Consulta el elemento de la posicion i-ésima
        }
        
        return fullAttack;
    }
    
    /**
     * Calcula la defensa de los escudos
     * @return La defensa
     */
    protected float sumShield(){
        float fullProtection = 0;
        
        for (int i=0; i<shields.size(); ++i){
            fullProtection += shields.get(i).protect();
        }
        
        return fullProtection;
    }
    
    /**
     * Calcula la defensa total
     * @return la defensa
     */
    protected float defensiveEnergy(){
        float defenseEnergy  = super.getIntelligence() + this.sumShield();
        return defenseEnergy;
    }
    
    /**
     * Maneja un ataque
     * @param receivedAttack la fuerza del ataque
     * @return si ha perdido por el ataque
     */
    private boolean manageHit(float receivedAttack){

        float defense = defensiveEnergy();
        if(defense < receivedAttack){
            gotWounded();
            incConsecutiveHits();
        } else {
            resetHits();
        }
        boolean lose = false;
        if(consecutiveHits == HITS2LOSE || dead()){
            resetHits();
            lose = true;
        }
        return lose;
    }
    
    private void resetHits(){
        this.consecutiveHits = 0;
    }
    
    private void incConsecutiveHits(){
        this.consecutiveHits++;
    }
    
    
    
}

